src/ir/selector.cc

```diff
@@ -0,0 +1,29 @@
+#include "src/ir/selector.h"
+
+#include "absl/types/variant.h"
+
+namespace raksha::ir {
+
+// A callable struct used to dispatch to the specific selector's
+// implementation of ToString.
+struct GenericSelectorToStringFunction {
+ public:
+  template<class T>
+  std::string operator()(const T &specific_selector) const {
+    return specific_selector.ToString();
+  }
+};
+
+// Apply the callable ToString struct to the specific_selector_ variant to
+// get the result of the specific_selector_'s ToString.
+std::string Selector::ToString() const {
+  return absl::visit(GenericSelectorToStringFunction(), specific_selector_);
```

Comment:
Is it "better" to use absl::variant and absl::visit like you have here than to use inheritance polymorphism (i.e., define some class that all the specific types of selectors we have extend, and then just call toString (which would need to be virtual) on the specific_selector_)? Why? This is not a rhetorical question -- I am new to Abseil and just curious.

It's not easy for me to see that there is a performance gain by doing it this Abseil way: I would guess that the vtable lookup that inheritance would cause to happen would work similarly to visit anyway. Or is the point more to avoid problems that C++ has with inheritance in general (diamand problem.. others ? ) ?

Reply:
The reason to use `variant` in this case is not for performance reasons; I agree that performance gains would be minimal, if there were any at all.

The big reason why I'm using a variant here is to make it easy to answer where this `Selector` lives and for how long. Imagine that we're using inheritence-based polymorphism and are dealing with a specific selector implemented as an arbitrary base type `BaseSelector` deriving from `Selector`. In order to use inheritance and virtual function calls, I would have to have anything that wished to access this `BaseSelector` through a variable that may hold any `Selector` by pointing at the `BaseSelector` through a `Selector *` or `Selector &` pointer. This means that I'm introducing a burden upon myself and all future programmers to think about the lifetime of that pointer: how long does the pointed-to `Selector` live? Who owns it? Who may be pointing to it at any given time? Sometimes the answer to those questions are super easy, sometimes they are complicated, and sometimes they seem easy when they are complicated. But in many circumstances, a pointer is the right thing to use anyway, so we just use inheritance.

This is not one of those cases. At the current time, a `Selector` is created basically only to participate in a `SelectorAccessPath`. It is convenient to place it as a member of a `SelectorAccessPath` field of type `std::vector<Selector>`, whereupon `std::vector` manages all of the lifetime concerns for us. If we used inheritance, on the other hand, we'd have to make that field of type `std::vector<Selector *>` and handle the lifetime concerns ourselves. By using a `Selector` containing an `absl::variant`, we can just move `Selector`s around as values and completely ignore the lifetime concern.